posthog/hogql_queries/lifecycle_query_runner.py

```diff
@@ -0,0 +1,252 @@
+                            GROUP BY start_of_period, status
+                        )
+                        WHERE start_of_period <= dateTrunc({interval}, {date_to})
+                            AND start_of_period >= dateTrunc({interval}, {date_from})
+                        GROUP BY start_of_period, status
+                        ORDER BY start_of_period ASC
+                    )
+                    GROUP BY status
+                """,
+                placeholders,
+                timings=self.timings,
+            )
+        return lifecycle_query
+
+    def to_persons_query(self) -> str:
+        # TODO: add support for selecting and filtering by breakdowns
+        with self.timings.measure("persons_query"):
+            return parse_select(
+                """
+                SELECT
+                    person_id, start_of_period as breakdown_1, status as breakdown_2
+                FROM
+                    {events_query}
+                """,
+                placeholders={"events_query": self.events_query},
+            )
+
+    def run(self) -> LifecycleQueryResponse:
+        response = execute_hogql_query(
+            query_type="LifecycleQuery",
+            query=self.to_query(),
+            team=self.team,
+            timings=self.timings,
+        )
+
+        # TODO: can we move the data conversion part into the query as well? It would make it easier to swap
+        # e.g. the LifecycleQuery with HogQLQuery, while keeping the chart logic the same.
+
+        # ensure that the items are in a deterministic order
+        order = {"new": 1, "returning": 2, "resurrecting": 3, "dormant": 4}
+        results = sorted(response.results, key=lambda result: order.get(result[2], 5))
+
+        res = []
+        for val in results:
+            counts = val[1]
+            labels = [
+                item.strftime("%-d-%b-%Y{}".format(" %H:%M" if self.query_date_range.interval_name == "hour" else ""))
+                for item in val[0]
+            ]
+            days = [
+                item.strftime("%Y-%m-%d{}".format(" %H:%M:%S" if self.query_date_range.interval_name == "hour" else ""))
+                for item in val[0]
+            ]
+
+            label = "{} - {}".format("", val[2])  # entity.name
+            additional_values = {"label": label, "status": val[2]}
+            res.append(
+                {
+                    "data": [float(c) for c in counts],
+                    "count": float(sum(counts)),
+                    "labels": labels,
+                    "days": days,
+                    **additional_values,
+                }
+            )
+
+        return LifecycleQueryResponse(result=res, timings=response.timings)
+
+    @cached_property
+    def query_date_range(self):
+        return QueryDateRange(
+            date_range=self.query.dateRange, team=self.team, interval=self.query.interval, now=datetime.now()
+        )
+
+    @cached_property
+    def event_filter(self) -> ast.Expr:
```

Comment:
Some of the methods are getting quite long and hard to read. I'd love if we could split them out into separate functions e.g. `_append_date_range`, `_append_properties` etc. at some point.

Timings could then also be done as decorators to help readability.

Reply:
Agreed. For now, I wanted to keep everything tight and packed. When we add the trends query, it's likely going to be too big for one file, so we're going to have to refactor things further... including likely moving event filters to their own files, and splitting them further.